integration/v3_grpc_test.go

```diff
@@ -432,3 +432,205 @@ func TestV3WatchFromCurrentRevision(t *testing.T) {
 		clus.Terminate(t)
 	}
 }
+
+// TestV3WatchCancel tests Watch APIs cancellation.
+func TestV3WatchCancel(t *testing.T) {
+	tests := []struct { // send Put, WatchRequest, and then receive WatchResponse in order
+		preqs []*pb.PutRequest
+		wreqs []*pb.WatchRequest
+
+		wresps []*pb.WatchResponse
+	}{
+		// one key, one put, one watch, one cancellation
+		{
+			[]*pb.PutRequest{
+				nil,
+				{
+					Key:   []byte("foo"),
+					Value: []byte("bar"),
+				},
+				nil,
+				nil,
+			},
+			[]*pb.WatchRequest{
+				{
+					CreateRequest: &pb.WatchCreateRequest{Key: []byte("foo")},
+				},
+				nil,
+				{
+					CancelRequest: &pb.WatchCancelRequest{WatchId: 0},
+				},
+				nil,
+			},
+
+			[]*pb.WatchResponse{
+				{
+					Header:   &pb.ResponseHeader{Revision: 1},
+					Created:  true,
+					Canceled: false,
+				},
+				{
+					Header:   &pb.ResponseHeader{Revision: 2},
+					Created:  false,
+					Canceled: false,
+					Events: []*storagepb.Event{
+						{
+							Type: storagepb.PUT,
+							Kv:   &storagepb.KeyValue{Key: []byte("foo"), Value: []byte("bar"), CreateRevision: 2, ModRevision: 2, Version: 1},
+						},
+					},
+				},
+				nil,
+				{
+					Header:   &pb.ResponseHeader{Revision: 2},
+					Created:  false,
+					Canceled: true,
+				},
+			},
+		},
+		// no key, no put, 2 watch, 2 cancellation
+		{
+			[]*pb.PutRequest{
+				nil,
+				nil,
+				nil,
+				nil,
+			},
+			[]*pb.WatchRequest{
+				{
+					CreateRequest: &pb.WatchCreateRequest{Key: []byte("foo")},
+				},
+				{
+					CreateRequest: &pb.WatchCreateRequest{Key: []byte("foo")},
+				},
+				{
+					CancelRequest: &pb.WatchCancelRequest{WatchId: 0},
+				},
+				{
+					CancelRequest: &pb.WatchCancelRequest{WatchId: 0},
+				},
+			},
+
+			[]*pb.WatchResponse{
+				{
+					Header:   &pb.ResponseHeader{Revision: 1},
+					Created:  true,
+					Canceled: false,
+				},
+				{
+					Header:   &pb.ResponseHeader{Revision: 1},
+					Created:  true,
+					Canceled: false,
+				},
+				{
+					Header:   &pb.ResponseHeader{Revision: 1},
+					Created:  false,
+					Canceled: true,
+				},
+				{
+					Header:   &pb.ResponseHeader{Revision: 1},
+					Created:  false,
+					Canceled: true,
+				},
+			},
+		},
+	}
+
+	for i, tt := range tests {
+		clus := newClusterGRPC(t, &clusterConfig{size: 3})
+
+		kvc := pb.NewKVClient(clus.RandConn())
+		wAPI := pb.NewWatchClient(clus.RandConn())
+
+		wStream, err := wAPI.Watch(context.TODO())
+		if err != nil {
+			t.Fatalf("#%d: wAPI.Watch error: %v", i, err)
+		}
+
+		wcreatedN := 0
+		for _, wreq := range tt.wreqs {
+			if wreq == nil {
```

Comment:
I set up the tests, with each put, watch request, watch response having exact the same number of elements
so that I can run each action by index and match event by index. Do you have any suggestion? I need `nil` to fill
those slice. And if `wreq` is `nil` it doesn't send watch request.